Sources/NIOFileSystem/Internal/ParallelDirCopy.swift

```diff
@@ -0,0 +1,159 @@
+
+        // We ignore the result of yield in all cases because we are not implementing back pressure
+        // and cancellation is dealt with separately.
+        @Sendable func yield(_ contentsOf: [DirCopyItem]) {
+            _ = copyRequiredQueue.source.yield(contentsOf: contentsOf)
+        }
+
+        // Kick start the procees by enqueuing the root entry,
+        // the calling function already validated the root needed copying.
+        _ = copyRequiredQueue.source.yield(.toCopy(from: .init(path: sourcePath, type: .directory)!, to: destinationPath))
+
+        // The processing of the very first item (the root) will increment this,
+        // after than when it hits zero we've finished.
+        // This does not need to be a ManagedAtomic or similar because:
+        // - All maintenance of state is done in the withThrowingTaskGroup callback
+        // - All actual file system work is done by tasks created on the `taskGroup`
+        var activeDirCount = 0
+
+        // Despite there being no 'result' of each operation we cannot use a discarding task group
+        // because we use the 'drain results' queue as a concurrency limiting side effect.
+        try await withThrowingTaskGroup(of: Void.self) { taskGroup in
+
+            // Code handling each item to process on the current task
+            // Side Effects:
+            // - Updates activeDirCount and finishes the stream if required.
+            // - Either adds a single task on `taskGroup` or none.
+            // Returns true if it added a task, false otherwise.
+            func onNextItem(_ item: DirCopyItem) -> Bool {
+                switch item {
+                case .endOfDir:
+                    activeDirCount -= 1
+                    if activeDirCount == 0 {
+                        copyRequiredQueue.source.finish()
+                    }
+                    return false
+                case let .toCopy(from: from, to: to):
+                    if from.type == .directory {
+                        activeDirCount += 1
+                    }
+                    taskGroup.addTask {
+                        try await copySelfAndEnqueueChildren(
+                            from: from,
+                            to: to,
+                            yield: yield,
+                            shouldProceedAfterError: shouldProceedAfterError,
+                            shouldCopyItem: shouldCopyItem
+                        )
+                    }
+                    return true
+                }
+            }
+
+            let iter = copyRequiredQueue.sequence.makeAsyncIterator()
+
+            // inProgress counts the number of tasks we have added to the task group
+            // Get up to the maximum concurrency first.
+            // We haven't started monitoring for task completion, so inProgress is 'worst case'.
+            var inProgress = 0
+            while inProgress <= maxConcurrentOperations {
+                try Task.checkCancellation()
```

Comment:
We are calling `iter.next()` on the next line which will check for cancellation when needed